paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/AddPartitionCommitCallback.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.hive;
+
+import org.apache.paimon.catalog.Identifier;
+import org.apache.paimon.data.BinaryRow;
+import org.apache.paimon.manifest.ManifestCommittable;
+import org.apache.paimon.table.FileStoreTable;
+import org.apache.paimon.table.sink.CommitCallback;
+import org.apache.paimon.table.sink.CommitMessage;
+import org.apache.paimon.utils.PartitionPathUtils;
+import org.apache.paimon.utils.RowDataPartitionComputer;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+/**
+ * Add all partitions extracted from the list of {@link ManifestCommittable} to Hive metastore and
+ * ignore existing partitions.
+ */
+public class AddPartitionCommitCallback implements CommitCallback {
+
+    private final Identifier identifier;
+    private final RowDataPartitionComputer partitionComputer;
+
+    private final IMetaStoreClient client;
+    private final StorageDescriptor sd;
+
+    public AddPartitionCommitCallback(
+            Identifier identifier, FileStoreTable table, IMetaStoreClient client) {
+        this.identifier = identifier;
+        this.partitionComputer =
+                new RowDataPartitionComputer(
+                        table.coreOptions().partitionDefaultName(),
+                        table.schema().logicalPartitionType(),
+                        table.schema().partitionKeys().toArray(new String[0]));
+
+        this.client = client;
+        try {
+            this.sd =
+                    client.getTable(identifier.getDatabaseName(), identifier.getObjectName())
+                            .getSd();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void call(List<ManifestCommittable> committables) {
```

Comment:
How about adding  options that  add partition  by comparing watermark and parition time.